src/include/RADOS/RADOS.hpp

```diff
@@ -0,0 +1,612 @@
+#include <tuple>
+#include <string>
+#include <string_view>
+#include <type_traits>
+#include <variant>
+
+#include <boost/asio.hpp>
+
+#include <boost/container/flat_map.hpp>
+#include <boost/container/flat_set.hpp>
+#include <boost/uuid/uuid.hpp>
+
+
+// HATE. LET ME TELL YOU HOW MUCH I'VE COME TO HATE BOOST.SYSTEM SINCE
+// I BEGAN TO LIVE.
+
+#include <boost/system/error_code.hpp>
+
+#include "include/function2.hpp"
+
+// Needed for type erasure and template support. We can't really avoid
+// it.
+
+#include "common/async/completion.h"
+
+// These are needed for RGW, but in general as a 'shiny new interface'
+// we should try to use forward declarations and provide standard alternatives.
+
+#include "include/rados/rados_types.hpp"
+#include "include/buffer.h"
+#include "include/object.h"
+
+#include "common/ceph_time.h"
+
+#include "librados/ListObjectImpl.h"
+
+#ifndef RADOS_UNLEASHED_HPP
+#define RADOS_UNLEASHED_HPP
+
+class CephContext;
+
+namespace RADOS {
+class Object;
+}
+namespace std {
+template<>
+struct hash<RADOS::Object>;
+}
+
+namespace RADOS {
+
+class RADOS;
+
+// Exists mostly so that repeated operations on the same object don't
+// have to pay for the string copy to construct an object_t.
+
+class Object {
+  friend RADOS;
+  friend std::hash<Object>;
+
+public:
+  Object(std::string_view s);
+  Object(std::string&& s);
+  Object(const std::string& s);
+  ~Object();
+
+  Object(const Object& o);
+  Object& operator =(const Object& o);
+
+  Object(Object&& o) = delete;
+  Object& operator =(Object&& o) = delete;
+
+  operator std::string_view() const;
+
+  friend bool operator <(const Object& lhs, const Object& rhs);
+  friend bool operator <=(const Object& lhs, const Object& rhs);
+  friend bool operator >=(const Object& lhs, const Object& rhs);
+  friend bool operator >(const Object& lhs, const Object& rhs);
+
+  friend bool operator ==(const Object& lhs, const Object& rhs);
+  friend bool operator !=(const Object& lhs, const Object& rhs);
+
+private:
+
+  static constexpr std::size_t impl_size = 4 * 8;
+  std::aligned_storage_t<impl_size> impl;
+};
+
+// Not the same as the librados::IoCtx, but it does gather together
+// some of the same metadata. Since we're likely to do multiple
+// operations in the same pool or namespace, it doesn't make sense to
+// redo a bunch of lookups and string copies.
+
+struct IOContext {
+  friend RADOS;
+
+  IOContext();
+  explicit IOContext(std::int64_t pool);
+  IOContext(std::int64_t _pool, std::string_view _ns);
+  IOContext(std::int64_t _pool, std::string&& _ns);
+  ~IOContext();
+
+  IOContext(const IOContext& rhs);
+  IOContext& operator =(const IOContext& rhs);
+
+  IOContext(IOContext&& rhs);
+  IOContext& operator =(IOContext&& rhs);
+
+  std::int64_t pool() const;
+  void pool(std::int64_t _pool);
+
+  std::string_view ns() const;
+  void ns(std::string_view _ns);
+  void ns(std::string&& _ns);
+
+  // Because /some fool/ decided to disallow optional references,
```

Comment:
It seems fairly obvious to me that since std::optional<T&> is an optional containing a reference, .clear should clear it, as clear sets it back to an 'empty' value. That is, clear is a property of optional no matter what's inside it.

(That and I'd expect std::optional<const T> to allow both clearing and reassignment, just not mutation of the value IN the optional.)